PolyEngine/Core/Src/Collections/BitMask.cpp

```diff
@@ -0,0 +1,382 @@
+}
+
+BitMask& BitMask::operator|=(const BitMask& rhs)
+{
+	//Equal Dynarray sizes
+	if (BitList.GetSize() == rhs.BitList.GetSize())
+	{
+		for (size_t i = 0; i < BitList.GetSize(); i++)
+			BitList[i] |= rhs.BitList[i];
+
+		if (rhs.BitsNumber > BitsNumber)
+			BitsNumber = rhs.BitsNumber;
+
+		return *this;
+	}
+
+	else if (BitList.GetSize() > rhs.BitList.GetSize())
+	{
+		for (size_t i = 0; i < rhs.BitList.GetSize(); i++)
+			BitList[i] |= rhs.BitList[i];
+
+		return *this;
+	}
+
+	else if (BitList.GetSize() < rhs.BitList.GetSize())
+	{
+		for (size_t i = 0; i < BitList.GetSize(); i++)
+			BitList[i] |= rhs.BitList[i];
+
+		size_t oldBitListSize = BitList.GetSize();
+
+		for (size_t i = 0; i < rhs.BitList.GetSize() - oldBitListSize; i++)
+			BitList.PushBack(ZERO);
+
+		for (size_t i = oldBitListSize; i < BitList.GetSize(); i++)
+			BitList[i] = rhs.BitList[i];
+
+		BitsNumber = rhs.BitsNumber;
+		return *this;
+	}
+	return *this;
+}
+
+BitMask& BitMask::operator^=(const BitMask& rhs)
+{
+	if (BitList.GetSize() == rhs.BitList.GetSize())
+	{
+		for (size_t i = 0; i < BitList.GetSize(); i++)
+			BitList[i] ^= rhs.BitList[i];
+
+		if (rhs.BitsNumber > BitsNumber)
+			BitsNumber = rhs.BitsNumber;
+
+		return *this;
+	}
+	else if (BitList.GetSize() > rhs.BitList.GetSize())
+	{
+		for (size_t i = 0; i < rhs.BitList.GetSize(); i++)
+			BitList[i] ^= rhs.BitList[i];
+
+		return *this;
+	}
+	else if (BitList.GetSize() < rhs.BitList.GetSize())
+	{
+		for (size_t i = 0; i < BitList.GetSize(); i++)
+			BitList[i] ^= rhs.BitList[i];
+
+		size_t oldBitListSize = BitList.GetSize();
+
+		for (size_t i = 0; i < rhs.BitList.GetSize() - oldBitListSize; i++)
+			BitList.PushBack(ZERO);
+
+		for (size_t i = oldBitListSize; i < BitList.GetSize(); i++)
+			BitList[i] = rhs.BitList[i];
+
+		BitsNumber = rhs.BitsNumber;
+		return *this;
+	}
+	return *this;
+}
+
+BitMask& BitMask::operator&=(const BitMask& rhs)
+{
+
+	if (BitList.GetSize() == rhs.BitList.GetSize())
+	{
+		for (size_t i = 0; i < BitList.GetSize(); i++)
+			BitList[i] &= rhs.BitList[i];
+
+		if (rhs.BitsNumber > BitsNumber)
+			BitsNumber = rhs.BitsNumber;
+
+		return *this;
+	}
+	else if (BitList.GetSize() > rhs.BitList.GetSize())
+	{
+		for (size_t i = 0; i < rhs.BitList.GetSize(); i++)
+			BitList[i] &= rhs.BitList[i];
+
+		for (size_t i = rhs.BitList.GetSize(); i < BitList.GetSize(); i++)
+			BitList[i] = ZERO;
+		return *this;
+	}
+	else if (BitList.GetSize() < rhs.BitList.GetSize())
+	{
+		for (size_t i = 0; i < BitList.GetSize(); i++)
+			BitList[i] &= rhs.BitList[i];
+
+		size_t oldBitListSize = BitList.GetSize();
+
+		for (size_t i = 0; i < rhs.BitList.GetSize() - oldBitListSize; i++)
+			BitList.PushBack(ZERO);
+
+		for (size_t i = oldBitListSize; i < BitList.GetSize(); i++)
+			BitList[i] = ZERO;
+
+		BitsNumber = rhs.BitsNumber;
+		return *this;
+	}
+	return *this;
+}
+
+bool BitMask::operator==(const BitMask rhs) const
+{
+	if (BitsNumber != rhs.BitsNumber)
+		return false;
```

Comment:
Not necessarly, if one of the bitmasks is prefixed with a lot of zeroes this would still be true.
It would probably be helpfull to add method called `Trim()` which resizes BitMast to size == position of the leftmost bit set to 1. This way you can first trim, or get size after potential trim operation, and compare them to each other.